Add multi-release compilation support to JDT

laeubi opened the pull request:
Currently JDT (like other IDEs) lacks good support for compilation of so called [multi-release-jar](https://openjdk.org/jeps/238) setups. Maven has quite good support for this by setting a `${release}` property per source folder and [enable an output flag](https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#multiReleaseOutput) but often needs some ugly workarounds due to missing support in IDEs.

This is an attempt to bring multi-release-compilation to JDT to finally have this feature supported at least in one of the major IDEs and make this powerful feature more widely adopted.

In its current form it does the following:

1. One can set a `release` attribute in the classpath of a source folder, this can look like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
		<attributes>
			<attribute name="module" value="true"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="src" path="src"/>
	<classpathentry kind="src" path="src9">
		<attributes>
			<attribute name="release" value="9"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="output" path="bin"/>
</classpath>
```
2. if such attribute is found JDT enables "multi-release-output" as per specification to place the compiled output of this source folder to `<outputfolder>/META-INF/versions/<release>`

The next step would be to further pass the release option to the compiler itself.

laeubi commented:
I now also passed this down to the compiler, inspecting the generated class files shows that the `target` is correctly applied, next step would be to add a testcase that checks this.

laeubi commented:
I now also added a test-case to show the new feature, so I think this is ready for review.

iloveeclipse commented on org.eclipse.jdt.core.tests.builder/src/org/eclipse/jdt/core/tests/builder/MultiReleaseTests.java:
```diff
@@ -0,0 +1,60 @@
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.tests.util.Util;
+
+public class MultiReleaseTests extends BuilderTests{
```
Please add `MultiReleaseTests` to `org.eclipse.jdt.core.tests.builder.BuilderTests.getAllTestClasses()`.

iloveeclipse commented:
Do you plan to add UI support for this?

laeubi commented:
> Do you plan to add UI support for this?

Yes if this is available then next step would be to add UI in Classpath Configuration Page

laeubi commented:
@iloveeclipse tests are all green now can you review/merge?

iloveeclipse reviewed (changes requested):
I've reviewed the test only and commented on test code only because I've realized there are too many open questions.

Some concept parts are unclear to me and I believe before we will check other code changes we should have a clear picture of the proposed concept.

Is the description of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/265 ticket up-to-date with this PR? It doesn't look like it is, please provide there a full picture of proposed concept.

Topics to cover:
- effects on class output directories
- effects on extra compilation checks (errors/warnings) needed
- effects on inter-project compilation
- effects on indexer / search / call hierarchy / UI

For example, just few questions I had in mind:

- It **seem** to me https://openjdk.org/jeps/238 expects that the "release" contains exactly a subset of "default" sources and no new API / classes. They say `By this scheme, it is possible for versions of a class designed for a later Java platform release to override the version of that same class designed for an earlier Java platform release.` For me it means, in the IDE we shouldn't allow any class in the "release" directory that doesn't exist in "default" source directory? What is our take on it? If this is not the case, I assume classes from "release" source folders shouldn't be available for compilation of "default" source folders? Means, if you have class `ForRelease9`, it shouldn't be resolvable from "SomeInnocentClassForRelease8"? I don't see it is solved yet, at least incremental build allows to see all classes from "release" source folder in "default" source folder.
- Should the "release" source folder be compiled last, after "default" source folder? My assumption is that "release" source folder is just "mirroring" a subset of classes from default source folder, so in order to resolve all dependencies to the default sources the order matters?
- If there are multiple "release" source folders, are they compiled in some specific oder? Does it matter?
- Given projects A (multi-release 8/21) and B (Java 21, uses A), the "release 21" class version from A should be used to compile project B in the IDE. This seem to be not implemented yet? 
- What is about indexer? Which version do we see on index? Both? Only one?
- The "search" / "call hierarchy" etc resolve classes for single JDK release only, but they would need to resolve depending on the source folder where they were executed?
- The https://openjdk.org/jeps/238 "Classpath and modulepath" section explains that depending on where the multi-release jar is placed, different classes will be visible (because of exported packages per module). This need to be understood / implemented regarding multiple projects case like A/B above.

iloveeclipse commented:
Note: I'm not against the idea of supporting multi-release compilation in the IDE in general, so don't see my negative review as a sign that the idea will never be accepted, it just needs more work.

laeubi commented on org.eclipse.jdt.core.tests.builder/src/org/eclipse/jdt/core/tests/builder/MultiReleaseTests.java:
```diff
@@ -0,0 +1,60 @@
+		IPath defaultReleaseClass = projectPath.append("bin/p/MultiReleaseType.class");
+		IPath java9ReleaseClass = projectPath.append("bin/META-INF/versions/9/p/MultiReleaseType.class");
+		expectingPresenceOf(defaultReleaseClass);
+		expectingPresenceOf(java9ReleaseClass);
```
Thanks for the hint, I now added a proper assertion.

laeubi commented on org.eclipse.jdt.core.tests.builder/src/org/eclipse/jdt/core/tests/builder/MultiReleaseTests.java:
```diff
@@ -0,0 +1,60 @@
+		expectingPresenceOf(java9ReleaseClass);
+	}
+
+}
```
> I assume test tests project with Java 8 default? Or 21? 9? It is not clear. Assuming the test default is Java 8 and "release" package is 21, this seem to be against proposed concept where project default target is always higher then any "release" source folder target. The test should have both positive / negative checks for that.

I have no explicitly set the default to Java 1.8 and added a check that it is compiled with that java version.

laeubi replied:
> 1. Different source folders with same "release" attribute should compile properly all non conflicting classes in same "release" specific output directory.
> 2. Different source folders with same "release" attribute and same classes (== conflict!) should produce an error for duplicated classes.

I now added test case for both scenarios.

laeubi replied:
> Multi-project tests. Project A (multi-release), B, and C, B has A in classpath, C has A in module path, test should verify which class versions are accessible from B, C and which are used during compilation etc.

I now have added a simple testcase for a project using the MR type on its classpath. For the module-path I'm not very familiar how it would look like, this is probably better handled in a separate PR when we have "plain" MR support.

laeubi commented:
@iloveeclipse I now have hopefully addressed the concerns and added more tests. It would be great if you can review again so this can be considered for M2 or at least M3 and I can work on the UI parts.

iloveeclipse commented:
@laeubi : seeing the date of M3 set to 16th of May I cannot guarantee that, I'm very busy right now, and this is a feature not not a regression for which I have the highest priority.

@eclipse-jdt/eclipse-jdt-committers : could someone please take a look?

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/State.java:
```diff
@@ -62,7 +62,7 @@ public class State {
 Map<String, ReferenceCollection> references;
 // keyed by qualified type name "p1/p2/A", value is the project relative path which defines this type "src1/p1/p2/A.java"
-public Map<String, String> typeLocators;
+public Map<String, Map<Integer,String>> typeLocators;
```
Please update the comment defining the format of this map.

Wait a minute: this change would break org.eclipse.pde.api.tools!

laeubi replied:
> Wait a minute: this change would break org.eclipse.pde.api.tools!

We should update API tools then to adapt to this change of (internal) API.

laeubi replied:
Done

stephan-herrmann commented:
I pulled the change and played with it resulting in this test project: 
[MR24.zip](https://github.com/user-attachments/files/20067207/MR24.zip)

While class files are produced in the expected output structure, compilation produces several unexpected errors. What's more: the set of reported errors depends on how exactly the build is triggered:
* a clean build tells me that `src9/p/B.java` cannot see class `A` and  `src24/p/B.java` can see neither class `A` nor `C`
* editing `src9/p/B.java` triggers an incremental build that let's all errors vanish
* even when the Problems view is clean, the editor for `D.java` complains "The method m() is undefined for the type B"
* saving `D.java` makes that last message appear in the problems view, too.
* in the editor of `D.java` select `B` and press **F3**: navigation takes as to `src/p/B.java`, which is wrong.

From browsing the code, I couldn't find any concepts that would define visibility between source folders.

@laeubi do you have plans how such inter-folder references are supposed to be resolved? 
* if yes please explain
* otherwise we need to go back to the drawing board

laeubi commented:
@stephan-herrmann I'll look into your example, please note that some problems with code navigation are cause because JDT currently seem to not handle MR jars correct see:

- https://github.com/redhat-developer/vscode-java/issues/3319

Beside that, as explained I do not want to solve everything in this PR but allow to at least enable basic MR compilation support. So more things will likely be to discovered (and need to solved), but at least the basic things should work (and people use this usually in a basic way and not create complex inter-dependencies between packages).

laeubi commented:
@stephan-herrmann I have now added your testcase and can confirm that it fails to compile even though I would expects it working (even though adding package private members / public methods to a MR type `B` introduces the risk of breaking the contract of Multirelease jar).

I could need some hints on how it is supposed to work with crossreferences between different source folders, as I can see `A` is compiled and present in the output folder, something still prevents it from being found when compiling `src9` + `src24`, so any place where I should specifically look at or put a break point to?

stephan-herrmann commented:
> @stephan-herrmann I have now added your testcase and can confirm that it fails to compile even though I would expects it working (even though adding package private members / public methods to a MR type `B` introduces the risk of breaking the contract of Multirelease jar).
> 
> I could need some hints on how it is supposed to work with crossreferences between different source folders, as I can see `A` is compiled and present in the output folder, something still prevents it from being found when compiling `src9` + `src24`, so any place where I should specifically look at or put a break point to?

Which classes are found is initially determined by the `NameEnvironment`, see, e.g., `AbstractImageBuilder.nameEnvironment`. During compilation, the `Compiler` stores the internal representations (ast and/or bindings) in its `LookupEnvironment`, which is alive as long as the `Compiler` is alive. This implies that you cannot use the same `Compiler` instance for source folders with different visibility scopes.

For an example how these issues can be resolved see how each `AbstractImageBuilder` is created for a `CompilationGroup` (`MAIN` or `TEST`).  See in particular that `JavaBuilder` has two distinct name environments: `nameEnvironment` and `testNameEnvironment`.

Before you start adopting that strategy, it would be good to know, what is the expectation when a project has version dependent source folders as well as one or more test source folders: what should be the exact visibilities of each folder?

laeubi commented:
I see al about to compiled files are put into the environment before starting the compiles so I'm a bit confused the *name* is not found but maybe something is cleared between the steps and I find it sometimes hard to track down where the exact error is generated in the compiler.

My idea was to maybe not have an own NameEnvironment/builder but maybe it needs to be the case here will investigate.

> Before you start adopting that strategy, it would be good to know, what is the expectation when a project has version dependent source folders as well as one or more test source folders: what should be the exact visibilities of each folder?

At laest for the maven case I'm not aware that test sources are handled differently, as from the outside (that is the testable interface) all releases must look the same. So what one would do is run the same test for each release on the given JVM with the same compiled test classes.

laeubi commented:
@stephan-herrmann thanks for the hints, I finally sorted it out, it was the other way round than I initially thought.

Because the about to compiled names *are set* the compiler ignores to lookup these types from the compiled binary locations! So what was needed instead is to limit the names to the types currently compiled for the current target level!

Now your provided testcase passes without a problem!

laeubi commented:
I now squashed everything into one commit and rebased on master.

stephan-herrmann commented:
Thanks for the update.

> Now your provided testcase passes without a problem!

I still see the error on `b.m()` in `D.java`.
* It is not shown after clean build :heavy_check_mark: 
* It is shown in the editor :x:
* It is shown after incremental build when saving `D.java` - in this case it also appears in the Problems view :x:

For testing the editor issue you may take inspiration from `ReconcilerTests*`.  Incremental build can be added to `MultiReleaseTests` without much ado.

Looking at the case of incremental compilation in the debugger, I don't see how the correct version of `B.java` could possibly be selected when using the same NameEnvironment for all source folders.

laeubi commented:
> I still see the error on b.m() in `D.java`

I'm not sure what errors is there, the test-case passes without a problem, did I missed some detail there?

> It is shown in the editor

I'm not sure I can add editor test here?

> It is shown after incremental build when saving D.java - in this case it also appears in the Problems view
> Incremental build can be added to MultiReleaseTests without much ado.

Can you give an existing class that tests incremental build at the moment? I'll try to take a look then to enhance the `MultiReleaseTest` with incremental support.

> Looking at the case of incremental compilation in the debugger, I don't see how the correct version of B.java could possibly be selected when using the same NameEnvironment for all source folders.

As for all MR types they must be binary compatible for all public/visible part to the outside, selection of *any* B.java must be sufficient. But I'll take a deeper look once I got it reproduced in the test.

laeubi commented:
> Looking at the case of incremental compilation in the debugger, I don't see how the correct version of B.java could possibly be selected when using the same NameEnvironment for all source folders.

@stephan-herrmann I don not have a testcase yet for your described error, but I think I found a solution, the problem is that the type lookup did not take the release into account, so it depends on order maybe, instead I now sort it descending by release and only include those that are smaller or equal to the release, that way it works like in the runtime.

laeubi commented:
@stephan-herrmann I now added a poor-man incremental test by simply replace the content of the java file with some new stuff then trigger the incremental build for that project and check everything is error free.

stephan-herrmann commented:
> @stephan-herrmann @iloveeclipse can we probably target this for the next release when master opens? Or are there any concerns left now with the enhanced test-case?

Given the progress so far, the next release sounds feasible, but we're not done yet.

I haven't yet done in-depth code reviewing but that shouldn't stop you from addressing some more issues:
* problems observed in the editor. You might start by debugging this in a runtime workbench, for a junit see my hint above about ReconcilerTests*. This is essential as it might prove that changes in the builder are not sufficient.
* it would be good to already prepare and link the necessary follow-up fix for API tools.
* the feature should also be surfaced in jdt.ui
* also the combination of release-specific source folders and test source folders deserves further scrutiny.
* similar for other projects referencing a project with release-specific source folders.

stephan-herrmann commented:
> > problems observed in the editor. You might start by debugging this in a runtime workbench, for a junit see my hint above about ReconcilerTests*. This is essential as it might prove that changes in the builder are not sufficient.
> 
> I didn't found yet a good entry point for debugging Editor Errors, any hints here?

Have you tried to convert one of your tests into a reconciler test? Without seeing what you tried, it's difficult to guess what you might be missing.

>  Strange enough the error does not surface anywhere else, neither the file is marked with a red mark nor the problems view show anything.

A few words of explanation:

* Reconciling works on working copies (which may or may not have unsaved changes), not files
* Any problems detected during reconciling are signaled using a **ProblemAnnotation** not a **Marker**. By design Annotations (org.eclipse.jface.text.source.Annotation) are shown in the editor only.
* The true power of this concept lies in the capability to verify several working copies with unsaved changes against each other, i.e., a working copy can see the unsaved content of any other working copy (there may be more to it based on the concept of the working copy **owner**).

stephan-herrmann commented on org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/LookupEnvironment.java:
```diff
@@ -1189,6 +1181,28 @@ public PlainPackageBinding createPlainPackage(char[][] compoundName) {
+}
+
+private int getRelease() {
+	int release =  (int) Double.parseDouble(CompilerOptions.versionFromJdkLevel(this.globalOptions.targetJDK));
```
Apparently you didn't find any utils for encoding JDK versions as (human readable) int. Given that nobody needed such encoding until now, I would prefer to keep with one of the existing encodings, either `ClassFileConstants.JDK9` ... or `ClassFileConstants.MAJOR_VERSION_9` ...,  where the latter format can be obtained like this `(int) (this.globalOptions.targetJDK >> 16)`.

laeubi replied:
I now replaced the parsing with the suggested simple form. One needs a "human readable" int because that is the actual value used to lookup the correct multi-release folder (and the MR-spec is all talking about that value). Encoding it as something else would make this not only harder to debug but also requires converting the value to the "real" one on multiple places.

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathMultiDirectory.java:
```diff
@@ -33,14 +34,16 @@ public class ClasspathMultiDirectory extends ClasspathDirectory {
 	this.sourceFolder = sourceFolder;
 	this.inclusionPatterns = inclusionPatterns;
 	this.exclusionPatterns = exclusionPatterns;
+	this.release = release>8?release:IReleaseAwareNameEnvironment.DEFAULT_RELEASE;
```
let's avoid the magic number `8`, please.

laeubi replied:
Done!

laeubi commented:
> Have you tried to convert one of your tests into a reconciler test?

Not yet, at the moment I try to get a breakpoint at the location where the error I see in the editor is produced.

> Any problems detected during reconciling are signaled using a ProblemAnnotation not a Marker. 

Thanks `ProblemAnnotation` seems to be exactly what I was looking for!

laeubi commented:
@stephan-herrmann 

API tools change is now here:
- https://github.com/eclipse-pde/eclipse.pde/pull/1805

I also solved the problem with the reconciler so the next build should be green again including the new testcase.

@iloveeclipse Do you further plan to review this? The PR is currently blocked by your negative review, so you either want to review or dismiss it as stale/outdated.

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/util/messages.properties:
```diff
@@ -19,6 +19,7 @@
 ### JavaModel messages.
 
 ### hierarchy
+AbstractImageBuilder_mr_missmatch_main=Target release for {0} ({1}) is lower or equal to default release ({2}).
```
What is the meaning of "default" here?

laeubi replied:
The default release is the one of the project (e.g. the usual code uses Java 8 (=default) where there is some code targeting Java 11 in a multi-release (=non default) way).

laeubi replied:
Yes the JEP has the "advantage" that it only has to think about a final JAR where they can always assume everything as before and in contrast one has "versioned (class) entries", but this seems not suitable wording here. I came up with "default" because we have some folders that using the default project settings and others diverge in the aspect of the multi-release target.

Also in maven one has usually "default executions" of a mojo and then one configures additionally ones. In any case I would be open to suggestions how to better word this, the project should be quite clear because its an error marker on a project file.

stephan-herrmann reviewed (commented):
I completed one round of in-depth code reviewing.

Please don't be scared by the amount of comments. I'll give a more high-level view of where I think we are later this week.

stephan-herrmann commented:
During code review I also collected code locations that are not release-aware but maybe should. I'll comment on relevance and possible process in a later comment.

* callers of `INameEnvironment.findType(..)` or of methods in `IReleaseAwareNameEnvironment` that pass `DEFAULT_RELEASE`:
  * the lambdas being passed in the two invocations of  `LookupEnvironment.askForTypeFromModules()`
  * `LookupEnvironment.createPlainPackage()`
  * `ModuleUtil.ModuleAccumulatorEnvironment.findType(char[][], char[])`
  * `CodeSnippetEnvironment.findType(char[][])`
  * `CodeSnippetEnvironment.findType(char[], char[][])`
  * `EvaluationContext.evaluateImports(INameEnvironment, IRequestor, IProblemFactory)`
  * `IndexBasedHierarchyBuilder.buildForProject(JavaProject, ArrayList, ICompilationUnit[], HashSet, IProgressMonitor)`
  * `SearchableEnvironment.findTypeInModules(char[][], ModuleBinding)` -- used by completion
* Generally `FileSystem` is not release-aware. Can we argue that all uses of the batch compiler should be configured for MR building via the build system (like maven)?
  * There is a subclass `NameEnvironmentWithProgress` which relates to DOM, not batch compilation, and is not release aware :-/
* `JavaSearchNameEnvironment` 
* `NameLookup.findType(String, String, boolean, int, boolean, boolean, boolean, IProgressMonitor)`
   * has callers from Java model & completion 
* More aspects that maybe should be handled in a release-aware way:
  * Should we support different module-info.java in release-specific source folders? If so methods like the following need an update:
    * `LookupEnvironment.getModule()`
    * `o.e.j.internal.compiler.env.ICompilationUnit.getModule()`
  * should lookup of packages be release-aware? 
    * see e.g. `NameEnvironment.listPackages()`

I also wonder, if multi-release structure should also be expected in binary dependencies, for a random example see the block `IPackageFragmentRoot.K_BINARY` in `org.eclipse.jdt.core.dom.PackageBinding.getAnnotations()`.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/NameLookup.java:
```diff
@@ -804,6 +809,16 @@ public Answer findType(
+			}).filter(pr -> pr.release() <= release)
+					.sorted(Comparator.comparingInt(PackageFragmentWithRelease::release).reversed())
+					.map(PackageFragmentWithRelease::fragment)
+					.toArray(IPackageFragment[]::new);
```
I adjusted that even though the previous implementation was more semantically and usually only a very few unnecessary `PackageFragmentWithRelease` instances are actually created. I think its always a question of what outweigh a little overhead versus hard to understand code and this likely will not become a performance hotspot.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/BatchImageBuilder.java:
```diff
@@ -158,25 +158,26 @@ protected void cleanOutputFolders(boolean copyBack) throws CoreException {
-										resource.setDerived(true, null);
-										return resource.getType() != IResource.FILE;
+				if (visited.add(outputFolder)) {
+					if (outputFolder.exists()) {
```
Because output folders in a MR jar are effectively nested, so if one performs a full clean of the project an output folder might al ready be deleted (what could of course always happen, e.g. if deleted by the user). Deleting an non existent folder sadly triggers a stupid exception that the folder does not exits.

I now added a comment here.

laeubi replied:
Hm... I tried to improve it ... what also will trigger another build.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -203,13 +196,23 @@ private void computeClasspathLocations(
 				} else {
+					IClasspathAttribute[] extraAttributes = entry.getExtraAttributes();
+					Map<String, String> map = Arrays.stream(extraAttributes).collect(Collectors.toMap(IClasspathAttribute::getName, IClasspathAttribute::getValue));
+					String release = map.get(IClasspathAttribute.RELEASE);
```
Done.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -203,13 +196,23 @@ private void computeClasspathLocations(
 								entry.ignoreOptionalProblems(),
-								externalAnnotationPath);
+								externalAnnotationPath,
+								release==null?-1:Integer.parseInt(release));
```
Yes

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -588,6 +591,21 @@ private NameEnvironmentAnswer findClass(String qualifiedTypeName, char[] typeNam
 		relevantLocations = this.binaryLocations;
 	}
 	NameEnvironmentAnswer suggestedAnswer = null;
+	if (release >= IReleaseAwareNameEnvironment.FIRST_MULTI_RELEASE) {
```
It is relevant in all cases here independent of where the locations are from (either module or "classic")

stephan-herrmann replied:
OK, I didn't expect seeing `ClasspathMultiDirectory` here. You're right, so for those the filtering and sorting is relevant, even when regarded as binary locations.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -588,6 +591,21 @@ private NameEnvironmentAnswer findClass(String qualifiedTypeName, char[] typeNam
 	}
 	NameEnvironmentAnswer suggestedAnswer = null;
+	if (release >= IReleaseAwareNameEnvironment.FIRST_MULTI_RELEASE) {
+		//we must filter and sort any classpath directory here
```
> extending the code comment would be great

Done

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/State.java:
```diff
@@ -61,8 +61,8 @@ public class State {
-// keyed by qualified type name "p1/p2/A", value is the project relative path which defines this type "src1/p1/p2/A.java"
-public Map<String, String> typeLocators;
+// keyed by qualified type name "p1/p2/A", value is a map that maps a release to the project relative path which defines this type "src1/p1/p2/A.java"
+public Map<String, Map<Integer,String>> typeLocators;
```
Maybe we can agree on:

1. The structure is only written once
2. It is read multiple times
3. java should be quite good on hash lookups of maps that in the case of no multi-release only ever contain one entry

If that is not satisfying, I could think about replacing the plain `Map` structure by a dedicated datastructure `TypeLocator` that optimize this particular lookup, but I must confess it feels hard to really proof what will be the performance improvement given that compilation often require a lot of String parsing and I/O access that likely is a much bigger penalty.

laeubi replied:
I now added a dedicated `TypeLocator` class that handles the default case with a map like before and only uses a multi-map in case a release is actually used.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/State.java:
```diff
@@ -180,17 +180,20 @@ StringSet getStructurallyChangedTypes(State prereqState) {
-		Set<Entry<String, String>> keyTable = this.typeLocators.entrySet();
-		for (Entry<String, String> entry : keyTable) {
-			String packageName = entry.getKey(); // is a type name of the form p1/p2/A
+		Set<String> keyTable = this.typeLocators.keySet();
```
Renamed it to `keySet`

laeubi commented:
@stephan-herrmann thanks for the detailed review, I added some new commits now to account for the comments, regarding your last comment see below:

> During code review I also collected code locations that are not release-aware but maybe should.

That's absolutely possible, currently JDT has some limited support for MR-Jars (mainly on the classpath of project/compiler) but there are some gaps (see for example also https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3168 and https://github.com/redhat-developer/vscode-java/issues/3319 ) but the most limiting one is the compilation part because it makes fixing other things less interesting. That's why I try to focus on that part first and if we have this the others will hopefully become prominent enough to be improved as well.

> Generally FileSystem is not release-aware. Can we argue that all uses of the batch compiler should be configured for MR building via the build system (like maven)? 

Usually that's the case (even though I have not looked at the particular case of `FileSystem` yet), e.g. Tycho already supports MR-Jar compilation similar to maven-compiler-plugin and also `javac` do not seem to have any MR support and relies on the tool to call it multiple times with different inputs (e.g. adding already compiled lower releases)

> Should we support different module-info.java in release-specific source folders? 

Basically yes, and even often a `module-info.class` is placed in the `META-INF/9` folder to hide it from other tools if one target Java 8. But that whole module story is something I rather would like to delay to a point where it becomes relevant, currently I focus on the non-modular use-case (also `module-info.class` often is generated afterwards in the build process).

> should lookup of packages be release-aware? 

I can't think about a case where it would be useful, and it seems to not causing any issues at the moment.

> I also wonder, if multi-release structure should also be expected in binary dependencies

Yes I think that is something complained [here](https://github.com/redhat-developer/vscode-java/issues/3319) but something one could solve as a separate issue once the `MultiReleaseAware` things are there.

laeubi commented:
Now one test fails with the stupid

> The type java.lang.Object cannot be resolved. It is indirectly referenced from required .class files

errors one sometimes see in the IDE as well, I'll try to take a look how this could relate to the performance optimization, but I think the review can still go on.

stephan-herrmann commented:
@laeubi I understand your approach of addressing plain compilation first, and postponing other topics. OTOH, a configuration that works during compilation but isn't fully supported, e.g., in search or code assist or any such, may give a negative impression of the IDE.

To not block the compiler / builder work, I could think of a simple disclaimer in the UI shown when selecting the "Release" property of a source folder. Something in the vein of "Projekt layout for creating a multi-release jar is an incubating feature, that may not be fully supported by all IDE features".

That would give us peace of mind for anything that's still work in progress (and were we cannot guarantee completion before the next release).

laeubi commented:
Its not working that bad, anyways I have applied your suggestion now as

![grafik](https://github.com/user-attachments/assets/2bf6cd70-48b4-447d-934b-e1d351c703af)

I also ran the tests locally and they all pass, maybe you can restart the Jenkins to see if its a timing problem? At worst we would need some kind of rebase because some other component conflicts here :-(

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CancelableNameEnvironment.java:
```diff
@@ -60,14 +60,26 @@ public NameEnvironmentAnswer findType(char[][] compoundTypeName) {
 	@Override
 	public NameEnvironmentAnswer findType(char[] typeName, char[][] packageName, boolean searchWithSecondaryTypes, char[] moduleName) {
-		return findType(typeName, packageName, moduleName);
+		return super.findType(typeName, packageName, moduleName);
```
OK, I agree this is more tricky then expected. In this case I would find it clearer to say:
```java
	@Override
	public NameEnvironmentAnswer findType(char[] typeName, char[][] packageName, boolean searchWithSecondaryTypes, char[] moduleName) {
		return findType(typeName, packageName, moduleName, NO_RELEASE);
	}
```
This would make it easier to see that `checkCanceled()` is indeed invoked. F3 would take you directly down 15 lines to the sibling method in this class.

Whether inheritance is to be preferred or delegation is a matter of taste, in particular in a language that has no support for implicit delegation. So let's keep the inheritance, please :)

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/State.java:
```diff
@@ -62,14 +70,12 @@ public class State {
 Map<String, ReferenceCollection> references;
 // keyed by qualified type name "p1/p2/A", value is a map that maps a release to the project relative path which defines this type "src1/p1/p2/A.java"
-public Map<String, Map<Integer,String>> typeLocators;
+public TypeLocator typeLocators;
```
Thanks for going the extra mile.

As a final polish I suggest to rename the class to `TypeLocators` (matching also the variable name), to indicate that one instance of that class manages many `String typeLocator` instances.

In that some round also the comment above the field should be updated / perhaps moved to the new class.

laeubi commented:
> > Its not working that bad,
> 
> no offense implied :)

I more mean that the rest of the IDE is not working that bad that I would say it gives a bad impression (especially compared to the previous state where it is not possible at all or only with workarounds compared to other IDEs), e.g. if I search for a Type JDT just list all MR-Types and the user has to choose what is of interest here and I don't have an immediate idea how to make it better without major change in the UI (e.g. using a tree instead of a list), but of course some glitches will be more visible now when we have this.

Still if adding some notes somewhere makes people more comfortable its not that bad either.

laeubi commented:
I now found the issue with the testfailure, it was introduced by in-lining the filtering of invalid releases, should be fixed now that the next build will be fine again!

stephan-herrmann commented:
As I'm done with in-depth code reviews, I'm fine with squashing and / or rebasing the existing changes, as long as additional changes come in their own separate commits :)

stephan-herrmann commented:
Here is my naive attempt at using this feature: 
[MRProject_release_option.zip](https://github.com/user-attachments/files/20884376/MRProject_release_option.zip)

In this variant (using `--release`) it looks broken in several regards.

The other problem can be observed if replacing only `.settings/org.eclipse.jdt.core.prefs` with
```
eclipse.preferences.version=1
org.eclipse.jdt.core.compiler.codegen.methodParameters=do not generate
org.eclipse.jdt.core.compiler.codegen.targetPlatform=9
org.eclipse.jdt.core.compiler.codegen.unusedLocal=preserve
org.eclipse.jdt.core.compiler.compliance=9
org.eclipse.jdt.core.compiler.debug.lineNumber=generate
org.eclipse.jdt.core.compiler.debug.localVariable=generate
org.eclipse.jdt.core.compiler.debug.sourceFile=generate
org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
org.eclipse.jdt.core.compiler.problem.reportPreviewFeatures=warning
org.eclipse.jdt.core.compiler.release=disabled
org.eclipse.jdt.core.compiler.source=9
```

However, the surface problem seems to depend on some caching, right now the above settings do not trigger an error, but in the debugger I can reproducibly see that `org.eclipse.jdt.internal.core.builder.NameEnvironment.findClass(String, char[], LookupStrategy, String, int)` is always called with `release` == `-1`.

stephan-herrmann commented:
I'm struggling to reproduce the problem with 1.8 without --release, so _perhaps_ this was a red herring.

Anyway I suggest we coordinate:
* what is the expected project configuration?
* what errors do we expect the compiler to find?

My naive understanding is that MRJARs serve the purpose to provide different variants that use the most appropriate API from the JDK depending on the target release version. So I think users should get help in making sure each source folder accesses only API that exists in that target version, right?

This goal I don't see achieved yet, and I haven't found any test that uses JDK API in a version dependent way.

On the long run, users might even expect to be informed when classes at higher version are not "compatible" with lower version ones, because this could cause NoSuchMethodError at runtime.

stephan-herrmann commented:
Fact is, today I can create a "error-free" MR Jar using this patch, which will throw NoSuchMethodError at runtime for either of these reasons:
* calling a method on my own class which doesn't exist in the version that is selected at runtime
* code from a folder specifically targeting release X calls a method that doesn't exist in JDK X

No future evolution needed for either of these.

stephan-herrmann commented:
> > what is the expected project configuration?
> 
> You have a project with a given "project default" target level (e.g. 8) your code is compatible with. Then you have one or more optimized variants (for whatever reason) that are requiring a higher target level (e.g. 11)

Which JDK should the project reference? Should it enable `--release` or not?

stephan-herrmann commented:
> > On the long run, users might even expect to be informed when classes at higher version are not "compatible" with lower version ones, because this could cause NoSuchMethodError at runtime.
> 
> I think that's nothing different than what we have today, if I compile my code e.g against a JDK21 preview API and later that is changed my jar needs to be updated. The same for any deprecated methods removed in a later JDK version of java.

I think you misunderstood my comment. I was thinking of a method call *within your own code* which at runtime gets the target class swapped for a variant that doesn't have the method. Should I send an example?

laeubi commented:
> Which JDK should the project reference? Should it enable `--release` or not?

The project must always use the highest compiled JDK release level, I do not see a reason why  `--release` should be forbidden.

laeubi commented:
> I was thinking perhaps you meant the MRJARs should always use --release

Maybe this should be the case, I don't know if there are other combinations where you can get with a JDK24 Java 8 byte code, if yes that should be supported even though it is very unusual. I only have seen projects using the `--release` option for it.

stephan-herrmann commented:
> > I was thinking perhaps you meant the MRJARs should always use --release
> 
> Maybe this should be the case, I don't know if there are other combinations where you can get with a JDK24 Java 8 byte code, if yes that should be supported even though it is very unusual. I only have seen projects using the `--release` option for it.

`--release` is not about getting the right kind of bytecode, it's about compiling against the right API.

laeubi commented:
> Here is

I'm looking into it now ...

> make a whitespace change to trigger incremental build, to get the project error-free

I think here it is where thing go wrong first, this should give an error always like after project clean as the Type `D` is not accessible at the default project code.

> Run with Java 9

So this is kind of expected, see above, will need a new testcase for this.

> Run with Java 17

This seems a bit expected here as the project is not using `org.eclipse.jdt.core.compiler.release=enabled`, so it is the same behavior I get with a plain JDT project when using a Java 24 VM and disable that option. What makes me think one probably should force `org.eclipse.jdt.core.compiler.release=enabled` for all multi-release enabled folders (and most likely one wants to use that for the main project sources as well but that's of course the choice of the user).

> Run with Java 24

This clearly violates the "binary compatible API contract" here and I'm not sure the compiler can detect this situation without complex analysis of all existing types and call sites. I would rather not implement this at this stage, of course such in deep analysis would be a great benefit to the user but that's more something one would perform outside compilation and given that depending on how it is used it might or might not work (a bit similar like null analysis where the compiler gives a hint a value *might* be `null`).

laeubi commented:
I now added a testcase for the case where the project settings are disabled the `target` option but for some cases changes in this are not taken into account. If I change the project settings to enable it before hands, this is actually working and I get the desired error. Looking through references of `CompilerOptions.target` only gives references in get/setMap and resetDefaults, so it seems I'm probably looking at the wrong item here... or something is cached before I can change it :-\

laeubi commented:
@stephan-herrmann I only found that `ClasspathMultiReleaseJar` uses the compliance option and though it might be evaluated during compilation but that's not the case instead it is determined when the classpath is first build so it does not change anything. Same for the `target` option and `ClasspathJrtWithReleaseOption` ... so long story short, I need to make `ClasspathLocation` itself release aware. What probably is a good thing because it is needed anyways and maybe even help in make one of the filtering obsolete. It might even help with https://github.com/redhat-developer/vscode-java/issues/3319 on the long run and might even contribute to https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3168 later on.

laeubi commented:
FYI @stephan-herrmann It is just a bit clumsy and needs more cleanup but I pushed it here just to show that in general it works as the test now passes. I enhanced the test in a way that it also check if in a higher java release one can use the method that is not available at the lower one.

Sadly the code is a little bit hard to extend I tried to make it more clear here:
- https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4143

but it seem to fail some one test, have to try it in smaller batches maybe.

What leads me to the question if you prefer I should try to make some code reorganization as separate PRs so they can be merged independently or keep them in the PR (that already has grown a bit).

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJrt.java:
```diff
@@ -149,10 +151,20 @@ public boolean equals(Object o) {
+	//TODO TEMP properly merge with ClasspathJrtWithReleaseOption
+	if (releaseNumber>= IReleaseAwareNameEnvironment.FIRST_MULTI_RELEASE) {
+		try {
+			 ClasspathJrtWithReleaseOption withReleaseOption = new ClasspathJrtWithReleaseOption(this.zipFilename, this.accessRuleSet,  this.externalAnnotationPath ==null?null:IPath.fromOSString(this.externalAnnotationPath), Integer.toString(releaseNumber));
```
I don't think we can afford to instantiate a `ClasspathJrtWithReleaseOption` for each lookup. To much caching inside that would go to waste.

laeubi replied:
Don't worry this was just a quick one to show it could work, now I need to make it not only work by being well integrated and nicely cached. With this change at least your described case with Java 17 using an API only available in Java 20+ will work (or more formally give an compile error) while it will be compile successful if you use it in a Java 20+ release folder.

laeubi replied:
@stephan-herrmann I now was able to demangle the `ClasspathJrtWithReleaseOption` and the `ClasspathJrt` to support a generic (cached) lookup of release based classe where the first one now is more a lightweight redirect to the super classes.

stephan-herrmann commented:
@laeubi I have been surprised by your progress so far on a path that I hadn't expected.

Unfortunately, I'm not sure that path will lead to success, because it doesn't seem to go well with the existing strategy of data caching.
* when building a given project we initialize all classpath locations so that lookup can rely on optimal cached data
  * during this process a `ClasspathJrtWithReleaseOption` is initialized for one specific release version
* now that we agree that different release folders need different release views of the JDK, this boils down to separate instances of `ClasspathJrtWithReleaseOption`

I don't see how those multiple instances of `ClasspathJrtWithReleaseOption` could be harmonized with the overall builder design.

We do have a precedent for multiple builder invocations with different configurations (see `testImageBuilder` in `JavaBuilder.buildAll()`). We do not have a precedent for changing the configuration during one builder invocation. Maybe there's a reason for that?

laeubi commented:
@stephan-herrmann actually I'm quite confident that this all will work out and haven't come to a conclusion that it is the wrong way its more feels like every test-case and the corresponding fix just reveals one more piece of the puzzle to be solved.

As I haven't worked with all these JDT internal before starting this adventure I still need to learn many things but I'm already quite satisfied by the result as already much more things got supported that will help developers that previously was more some kind of good-luck and a good test suite with different JDKs.

One thing about performance:
I have tried to my best to keep the impact for the common case (no multi-release) negligible for the price of having maybe a little less performance in the case of Multi-Release compile because this is a very rare case and if you do it correctness is more important than performance, don't expect people are writing large parts of their code in a multi-release way that's simply not something you want to do, I think [the spec explain it quite good](https://openjdk.org/jeps/238):

> Library and framework maintainers can gradually migrate to and support new features while still carrying around support for the old features

So the ultimate goal would always be to keep the amount of code that is MR as possible.

laeubi commented:
It seems it is working way to good now and some test fail because they compile against latest Java API by not using the target option, will add an additional check for that...

laeubi commented:
In the end the solution seems quite obvious ... instead of trying to figure out the release after the fact, use a `NameEnvironment` per release and let it carry the target release itself. This seems like a middle ground between what @stephan-herrmann suggested to use a complete own compilation groups (`MAIN` vs `TEST`) and what I used here initially. Unfortunately some of the interesting work I done here to allow lookups based on a release parameter are now likely not needed anymore (for this feature) and can be handled much more straight forward up in the chain and even with less influence on the standard compilation case.

Because of this, I now have made [a backup of my branch](https://github.com/laeubi/eclipse.jdt.core/tree/add_multi_release_support_iteration1) and will revert / rework some of the parts, the good thing is that on the way there where added some very good test-cases now so that changes will be easier to test and verify.

If all works well I plan to squash all changes into a _second commit_ and rebase it on master for a next round of review so one can easily see what changed compared to the first in-deep review.

laeubi commented:
@stephan-herrmann with these changes I now was able to get rid of the releaseaware environment for the compilation path entirely so the changes here are now quite minimal.

For the reconciler case it is a bit trickier, I'm currently toying around a bit to choose a similar approach for the `NameLookup` ...

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java:
```diff
@@ -427,18 +433,74 @@ protected void compile(SourceFile[] units, SourceFile[] additionalUnits, boolean
+					this.compiler.options.release = true;
+					try {
+						this.compiler.lookupEnvironment.nameEnvironment = this.javaBuilder.getNameEnvironment(release);
+					} catch (CoreException e) {
+						List<IContainer> list = entry.getValue().stream().map(sf -> sf.sourceLocation.sourceFolder)
+								.distinct().toList();
+						for (IContainer container : list) {
+							createProblemFor(container, null,
+									NLS.bind(Messages.AbstractImageBuilder_env_failed,
+											new Object[] { release, e.getStatus()}),
```
thanks for trying to handle this exception, but does it really signal something that warrants a problem marker? Certainly "LookupEnvironment" is not a concept that should be mentioned in user-facing messages.
Unfortunately, the location where the exception is raised does not give any hint on why it can occur.

laeubi replied:
`org.eclipse.jdt.internal.core.builder.JavaBuilder.initializeBuilder(int, boolean)` init the name environment that then is shared with the `AbstractImageBuilder` ... I could init them all there as well but though it might be better to be a bit more lazy here.

stephan-herrmann replied:
OK, throwing `ImageBuilderInternalException` makes exception handling look more like existing code, but I don't think that the exception thrown here will be caught somewhere in every possible call path. E.g., we may be indirectly called from `IncrementalImageBuilder.incrementalBuildLoop()` where I don't see any catching of this particular exception in place. Since it's a RuntimeException it's your responsibility to ensure proper catching :)

To me it would look safer if the exception is indeed handled here, but in a way that re-uses or copies the code combining logging and a specific buildpath error marker.

laeubi replied:
I found at least three different prior-arts in this class ... seems I decided for the wrong one, I now used your recommended approach.

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java:
```diff
@@ -427,18 +433,74 @@ protected void compile(SourceFile[] units, SourceFile[] additionalUnits, boolean
+								.distinct().toList();
+						for (IContainer container : list) {
+							createProblemFor(container, null,
+									NLS.bind(Messages.AbstractImageBuilder_mr_missmatch_main,
```
Did I already suggest to expand "default release" to "the project's default release" in this message?

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java:
```diff
@@ -427,18 +433,74 @@ protected void compile(SourceFile[] units, SourceFile[] additionalUnits, boolean
+											new Object[] { release, e.getStatus()}),
+									JavaCore.ERROR);
+						}
+						continue;
```
so, when `CoreException` was thrown, nothing from this source folder is getting compiled? We should try really hard, to avoid that.

laeubi replied:
Yes it seems better to have an error marker on one source folder and let the user fix that than break everything in the middle and not compile even more things. If a fatal error occurs (see above) I'm not sue what one can do better than not proceed with compilation?

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/JavaBuilder.java:
```diff
@@ -715,6 +720,21 @@ private int initializeBuilder(int kind, boolean forBuild) throws CoreException {
+	if (this.releaseSpecificEnvironments == null) {
+		this.releaseSpecificEnvironments = new HashMap<>();
+	}
+	INameEnvironment environment = this.releaseSpecificEnvironments.get(release);
+	if (environment == null) {
+		environment = new NameEnvironment(this.workspaceRoot, this.javaProject, this.binaryLocationsPerProject,
+				this.notifier, CompilationGroup.MAIN, release);
+		this.releaseSpecificEnvironments.put(release, environment);
+	}
```
`computeIfAbsent()` ? :)

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -203,13 +216,25 @@ private void computeClasspathLocations(
 					}
 					bLocation.patchModuleName = patchedModuleName;
 				} else {
+					String release = Arrays.stream(entry.getExtraAttributes())
```
After seeing several new implementations of this lookup, may I suggest `org.eclipse.jdt.internal.core.ClasspathEntry.getExtraAttribute(IClasspathEntry, String)`? :)

laeubi replied:
As we have `org.eclipse.jdt.core.IClasspathEntry.isTest()` I could of course ad something similar for release as well, I just try to keep the API footprint low, just let me know and I'll add something.

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -203,13 +216,25 @@ private void computeClasspathLocations(
+						mustSortOutput = true;
+						finalOutputFolder = outputFolder.getFolder(new Path(String.format("META-INF/versions/%s", release))); //$NON-NLS-1$
+					}
+					int sourceTargetRelease = release==null?JavaProject.NO_RELEASE:Integer.parseInt(release);
```
I can't help, but `sourceTargetRelease` sounds contradictory to me.

laeubi replied:
renamed

stephan-herrmann commented on org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/ReconcilerMultiReleaseTests.java:
```diff
@@ -0,0 +1,173 @@
+						}
+					}
+				""");
+		mrproject.setOption("org.eclipse.jdt.core.compiler.release", "enabled");
```
please use constants

laeubi replied:
Done

stephan-herrmann reviewed (changes requested):
I completed another round of source level review. 
I still plan to do some more black box testing, next week, hopefully.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java:
```diff
@@ -427,18 +433,74 @@ protected void compile(SourceFile[] units, SourceFile[] additionalUnits, boolean
+					}
+					if (!oldRelease && !warnAboutMissingReleaseFlag) {
+							for (SourceFile source : entry.getValue()) {
+								IProject project = source.sourceLocation.sourceFolder.getProject();
```
Fixed

laeubi reviewed (commented):
I have now hopefully addressed the concerns!

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/SearchableEnvironment.java:
```diff
@@ -91,17 +91,19 @@ public class SearchableEnvironment
 	 * Creates a SearchableEnvironment on the given project
 	 */
-	public SearchableEnvironment(JavaProject project, org.eclipse.jdt.core.ICompilationUnit[] workingCopies, boolean excludeTestCode) throws JavaModelException {
+	public SearchableEnvironment(JavaProject project, org.eclipse.jdt.core.ICompilationUnit[] workingCopies, boolean excludeTestCode, int release) throws JavaModelException {
```
It's the boundary between relevant and not relevant that looks fuzzy to me. Should all this be covered by the disclaimer "IDE functions may or may not respect this setting"? Or can you give some sort of guarantee like " is respected in all modes of building" ..?

laeubi replied:
Basically relevant for me is at the moment the case in the Editor where I edit some code and should get the same result as if I do a standard compile because as a user I can not really distinguish between a "real" clean compile and an incremental one in the reconciler case.

But `SearchableEnvironment` is used (or created) in multiple other places, e.g. content assist, hierarchy or search. These are of course not irrelevant to behave better/more correct but this is almost status-quo and I would touch those places better afterwards. e.g. https://github.com/redhat-developer/vscode-java/issues/3319 is a good example where one wants better support and where passing a release to the search environment might help. But I have not investigated this further.

laeubi commented:
I have now cleared up the remaining comments and the build should be more stable this week hopefully :-)

laeubi commented:
It was just due to changed text in messages, so next build should be green.

@stephan-herrmann do you like to review something from your current comments or is anything addressed? If yes then I'd like to rebase and squash again to be ready for the next round, I'm quite confident we have now a quite usable state with already some useful validations in place.

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/SearchableEnvironment.java:
```diff
@@ -91,17 +91,19 @@ public class SearchableEnvironment
 	private long timeSpentInFindTypes;
 
 	private List<IPackageFragmentRoot> unnamedModulePackageFragmentRoots;
+	private int release;
```
> Maybe you could find a central location where the core concepts of MR builds are explained and then link to that from locations like this one?

Searching for a central location for overview documentation seems to be a recurring topic :)

Here's an attempt at explaining a bit more what's relevant right here:
```
When {@link release} is not {@link JavaProject#NO_RELEASE} then this SearchableEnvironment
will define a view that prefers to search in locations that best match the given {@link release}
(priority on exact version match, then going down to older release versions until a type is found).
```

BTW: if `NO_RELEASE` is passed, will it search _in all locations_ or only in the locations of the _project default release_?

stephan-herrmann replied:
Given the cross-cutting nature of this change a central location for overview documentation would still be great, but I checked that browsing the call hierarchy of `IClasspathAttribute.RELEASE` brings us close to relevant locations.

laeubi commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/NameEnvironment.java:
```diff
@@ -441,7 +458,15 @@ private void computeClasspathLocations(
 		}
 	}
-
+	//sort according to release
```
I have now enhanced the description, I hope it makes it more clear. About a general description I'm not sure if it is needed and if where to place it. The closest would be `ClasspathMultiReleaseJar` and `ClasspathJrtWithReleaseOption` but it does not has any documentation at all.

stephan-herrmann replied:
Thanks, ok, except for the question ascending vs descending (see other comment).

stephan-herrmann commented:
I outsourced one observation to new issue #4268. Since that one affects compilation, I think it should be addressed even before diving into other IDE functions (search etc.).

laeubi commented:
First of all @stephan-herrmann many thanks for reviewing this and all the recommendations.

> squash and rebase everything

Done

> confirm that you are ready to merge the necessary changes in PDE. 

I have updated the PR here:
- https://github.com/eclipse-pde/eclipse.pde/pull/1805

> How can we get both changes in for the same I-build?

It will of course not compile with the local verification build so in this special case we need to merge _both_ PRs, start a new I-Build and crossing fingers. Then once the new build repo is available and shows some issues we can fix them using the regular process.

laeubi commented:
> Perhaps you'd even ask for an extra I-build just for safety?

Started one here:
- https://ci.eclipse.org/releng/job/Builds/job/I-build-4.37/90/

laeubi commented:
The I-Build succeeded :1st_place_medal: 

Eclipse downloads:
https://download.eclipse.org/eclipse/downloads/drops4/I20250731-0550

Build logs and/or test results (eventually):
https://download.eclipse.org/eclipse/downloads/drops4/I20250731-0550/testResults.php

Software site repository:
https://download.eclipse.org/eclipse/updates/4.37-I-builds

Specific (simple) site repository:
https://download.eclipse.org/eclipse/updates/4.37-I-builds/I20250731-0550

Equinox downloads:
https://download.eclipse.org/equinox/drops/I20250731-0550
